internal/jimm/juju/model.go

```diff
@@ -728,52 +741,81 @@ func (j *JujuManager) ListModels(ctx context.Context, user *openfga.User) ([]bas
 	}
 
 	// Create map for lookup later
-	modelsMap := make(map[string]dbmodel.Model)
+	modelsMap := make(map[uint][]dbmodel.Model)
 	// Find the controllers these models reside on and remove duplicates
 	var controllers []dbmodel.Controller
 	seen := make(map[uint]bool)
 	for _, model := range models {
-		modelsMap[model.UUID.String] = model // Set map for lookup
+		models := modelsMap[model.ControllerID]
+		if models == nil {
+			models = []dbmodel.Model{}
+		}
+		models = append(models, model)
+		modelsMap[model.ControllerID] = models
+
 		if seen[model.ControllerID] {
 			continue
 		}
 		seen[model.ControllerID] = true
 		controllers = append(controllers, model.Controller)
 	}
 
-	// Call controllers for their models. We always call as admin, and we're
-	// filtering ourselves. We do this rather than send the user to be 100%
-	// certain that the models do belong to user according to OpenFGA. We could
-	// in theory rely on Juju correctly returning the models (by owner), but this
-	// is more reliable.
+	// The call to ListModels is currently broken in Juju because it always consults
+	// mongo for authorization. Because of that we have to resort to ModelInfo calls.
 	var userModels []base.UserModel
 	var mutex sync.Mutex
-	err = j.forEachController(ctx, controllers, func(_ *dbmodel.Controller, api API) error {
-		ums, err := api.ListModels(ctx)
-		if err != nil {
-			return err
-		}
-		mutex.Lock()
-		defer mutex.Unlock()
-
-		// Filter the models returned according to the uuids
-		// returned from OpenFGA for read access.
-		//
-		// NOTE: Controller models are not included because we never relate
-		// controller models to users, and as such, they will not appear in the
-		// authorised uuid map.
-		for _, um := range ums {
-			mapModel, ok := modelsMap[um.UUID]
-			if !ok {
-				continue
+	eg := new(errgroup.Group)
+	for _, ctl := range controllers {
+		eg.Go(func() error {
```

Comment:
Can you take a look at declaring this function a few lines above so that it's less nested or making it a true function.

Reply:
done